NETOBSERV-1259: Adding Netobserv DNS tracking blog (#53)
netobserv/documents: blogs/dns_tracking/dns_tracking.md

# Network Observability Per Flow DNS tracking

![logo](./images/dns_tracking_logo.png)

By: Julien Pinsonneau, Mehul Modi and Mohamed S. Mahmoud

> <style>body {text-align: justify}</style>

Author: it looked good in my ide too

> <style>body {text-align: justify}</style>

Reviewer: Seems like it's not `md` formatting option, so it is visible when you display this as an md file.  The blog team recommends md formatting.

In today's interconnected digital landscape, Domain Name System (DNS) tracking
plays a crucial role in networking and security.
DNS resolution is a fundamental process that translates human-readable domain
names into IP addresses, enabling communication between devices and servers.

> However, this process also presents opportunities for monitoring and analysis,

Reviewer 2: Awesome ! I like that one :rocket: 
https://github.com/netobserv/documents/pull/53/commits/267b3b733cb84342bd042862820bcfcd03c40988

which can be achieved through innovative technologies like
eBPF (extended Berkeley Packet Filter).
In this blog post, we'll delve into the world of DNS tracking using eBPF
tracepoint hooks, exploring how this powerful combination can be used for
various purposes, including network monitoring and security enhancement.

## Understanding DNS Resolution

Before diving into the specifics of eBPF tracepoint hooks, let's briefly
recap how DNS resolution works.

> When you enter a website's domain name (e.g., www.example.com) in your

Reviewer: Can you figure out a way to remove the clickable link?  Maybe:
```suggestion
When you enter a website's domain name, such as www.example.com, in your
```

Reviewer 2: https://github.com/netobserv/documents/pull/53/commits/267b3b733cb84342bd042862820bcfcd03c40988

browser, your computer needs to find the corresponding IP address.
This process involves multiple steps, including querying DNS servers,
caching responses, obtaining the IP address to establish a connection

> and caching response for future re-occurrence of same DNS query.

Reviewer 3: ```suggestion
and caching response for future re-occurrence of the same DNS query.
```

## Utilizing Tracepoint Hooks for DNS Tracking

Tracepoint hooks are predefined points in the Linux kernel where eBPF
programs can be attached to capture and analyze specific events.
For DNS tracking, we leveraged tracepoint hooks associated with DNS

> resolution processes specifically `tracepoint/net/net_dev_queue` tracepoint,

Reviewer 3: ```suggestion
resolution processes, specifically the `tracepoint/net/net_dev_queue` tracepoint.
```

> Then we parse the DNS header to determine if its query or response, attempt

Reviewer 3: ```suggestion
Then we parse the DNS header to determine if it is a query or a response, attempt
```

> to correlate query and response for specific DNS transaction and then record the

Reviewer 3: ```suggestion
to correlate the query or response for specific DNS transaction, and then record the
```

Reviewer 3: In this list, I think it would be clearer to keep "or" consistent.

> to correlate the query or response for specific DNS transaction, and then record the

Reviewer: ```suggestion
to correlate the query or response with a specific DNS transaction, and then record the
```

Reviewer 2: https://github.com/netobserv/documents/pull/53/commits/267b3b733cb84342bd042862820bcfcd03c40988

elapsed time to compute DNS latency. Further, DNS network flows are enriched
to include fields, viz. DNS Id, DNS latency and DNS response codes to help
build graphs with aggregated DNS statistics and to help filtering on specific

> fields in Network Observability console.

Reviewer 3: ```suggestion
fields in the Network Observability console.
```

## Potential Use Cases

DNS tracking with eBPF tracepoint hooks can serve various purposes:

- Network Monitoring: Gain insights into DNS queries and responses,
helping network administrators identify unusual patterns,
potential bottlenecks, or performance issues.

- Security Analysis: Detect suspicious DNS activities, such as domain
name generation algorithms (DGA) used by malware,
or identify unauthorized DNS resolutions that might indicate a security breach.

- Troubleshooting: Debug DNS-related issues by tracing DNS resolution steps,
tracking latency, and identifying misconfigurations.

## How to enable DNS tracking

By default DNS tracking is disabled because it requires
`privileged` access, to enable this feature we need to create a flow

> collector object with the following fields enabled in eBPF config

Reviewer 3: ```suggestion
collector object with the following fields enabled in the eBPF config
```

> section

Reviewer: Do you expect readers to know how to access the eBPF config section?  If not, it would be helpful to add some details about how to access it.

Author: the eBPF config section is the following  yaml snippet in the same sentence, do u think we need to clarify it more ?

```yaml
apiVersion: flows.netobserv.io/v1beta1
kind: FlowCollector
metadata:
  name: cluster
spec:
  agent:
    type: EBPF
    ebpf:
      privileged: true
      features:
        - DnsTacking
```

## A quick tour in the UI

> Once `DnsTacking` feature enabled, the Console plugin will automatically

Reviewer 4: feature `is` enabled.

Author: based on the last diffs in the operator PR the feature is called `DNSTracking`

Reviewer 3: I add also: Once _the_ `DNSTracking` feature _is_ enabled

Reviewer: ```suggestion
Once the `DnsTracking` feature is enabled, the Console plugin will automatically
```

> adapt to provide additional filters and show informations across views.

Reviewer 3: ```suggestion
adapt to provide additional filters and show information across views.
```

Open your OCP Console and move to
`Administrator view` -> `Observe` -> `Network Traffic` page as usual.

> Three new filters, `DNS Id`, `DNS Latency` and `DNS Response Code` will be available
> in the common section:

Reviewer 3: ```suggestion
Three new filters, `DNS Id`, `DNS Latency` and `DNS Response Code` are available
in the common section:
```

Reviewer 3: Just a suggestion, but present tense is more appropriate for procedural information than future tense.

![dns filters](./images/dns_filters.png)

> The first one will allow you to filter on a specific DNS Id to correlate

Reviewer 3: ```suggestion
The first one allows you to filter on a specific `DNS Id` to correlate
```

with your query.

![dns id](./images/dns_id.png)

The second one helps to identify potential performance issues by looking at DNS
resolution latency.

![dns latency more than](./images/dns_latency_more_than.png)

The third filter surfaces DNS response codes, which can help detect errors or
unauthorized resolutions.

![dns rcode](./images/dns_response_code.png)

### Overview

> New graphs will be introduced in the `advanced options` -> `manage panels` popup:

Reviewer 3: ```suggestion
New graphs are introduced in the `advanced options` -> `manage panels` popup:
```

> New graphs are introduced in the `advanced options` -> `manage panels` popup:

Reviewer 2: https://github.com/netobserv/documents/pull/53/commits/267b3b733cb84342bd042862820bcfcd03c40988

![advanced options 1](./images/advanced_options1.png)

> - Top 5 average DNS latencies

Reviewer: As a critique on the dashboard itself, this is not a useful way to display DNS latency.  You probably want to see low, average, high.

- Top 5 DNS response code

> - Top 5 DNS response code stacked with total

Reviewer 4: Can be updated to Top X

Reviewer 2: sure https://github.com/netobserv/documents/pull/53/commits/56b52e776296548ec4e44f79729c4452f3aa2bf3

![dns graphs 1](./images/dns_graphs1.png)

> ![dns graphs 2](./images/dns_graphs2.png)

Reviewer: A critique on the dashboard itself - it should not include the "no error" in the totals.

Reviewer 2: that's optionnal and can be hidden using the kebab menu on top right corner :+1:

### Traffic flows

> The table view will get the new DNS columns `Id`, `Latency` and

Reviewer 3: ```suggestion
The table view adds the new DNS columns `Id`, `Latency` and
```

> `Response code` available from the `advanced options` -> `manage columns` popup

Reviewer 3: ```suggestion
`Response code`, which are available from the `advanced options` -> `manage columns` popup.
```

![advanced options 2](./images/advanced_options2.png)

> The DNS flows will show these informations in both table and side panel:

Reviewer 3: ```suggestion
The DNS flows display this information in both the table and the side panel:
```

> The DNS flows display this information in both the table and the side panel:

Reviewer: Note that this is where you would find the DNS Id to use in the DNS filters query above.

![dns table](./images/dns_table.png)

## Future support

- Adding tracking capability for mDNS

- Adding support for DNS over TCP

> - Investigate options to handle DNS over TLS where the DNS header is fully encrypted.

Reviewer 3: ```suggestion
- Investigating options to handle DNS over TLS where the DNS header is fully encrypted.
```

Reviewer 3: In lists, you want to keep the verbs the same

# Network Observability Per Flow DNS tracking

![logo](./images/dns_tracking_logo.png)

By: Julien Pinsonneau, Mehul Modi and Mohamed S. Mahmoud

In today's interconnected digital landscape, Domain Name System (DNS) tracking
plays a crucial role in networking and security. DNS resolution is a fundamental
process that translates human-readable domain names into IP addresses, enabling
communication between devices and servers. It can also be a security vulnerability
and a tool for firewalling and blocking access to content, depending on how it is
configured and manipulated. For operations excellence, DNS
resolution benefits from monitoring and analysis, which can be achieved through
innovative technologies like eBPF (extended Berkeley Packet Filter). In this
blog post, we'll delve into the world of DNS tracking using eBPF tracepoint
hooks, exploring how this powerful combination can be used for various purposes,
including network monitoring and security enhancement.

## Understanding DNS Resolution

Before diving into the specifics of eBPF tracepoint hooks, let's briefly recap

> how DNS resolution works. In a Kubernetes architecture, DNS (Domain Name System)

Reviewer 3: ```suggestion
how DNS resolution works. In a Kubernetes architecture, DNS
```

Reviewer 3: Since you already defined this acronym earlier, I don't think you need to define it again. In product docs, we typically only define for the first mention on the page then the rest of the mentions are acronym.

Reviewer 2: thanks ! https://github.com/netobserv/documents/pull/53/commits/29c84753d09a32904baa80885f7b9ab50896a4ee

# Network Observability Per Flow DNS tracking

<p align="center">
  <img src="dns_tracking_logo.png" alt="logo" width="25%"/>
</p>

By: Julien Pinsonneau, Mehul Modi and Mohamed S. Mahmoud

> In today's interconnected digital landscape, Domain Name System (DNS) tracking

Reviewer 5: I'd like to see more general information about network observability -- this is an opportunity to raise more awareness on the overall solution -- if we don't promote, we are missing the opportunity.  Even if we just talk about this being part of NOO and provide links to other information.  To me, this is less about this specific feature and more about how NOO has been enhanced now that we have it.
